exonum/src/helpers/fabric/details.rs

```diff
@@ -721,3 +766,97 @@ impl Command for GenerateTestnet {
+            GenerateNodeConfig::addresses(&ctx),
+            (external.to_string(), "0.0.0.0:1234".parse().unwrap())
+        );
+
+        let external = "127.0.0.1";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (
+                SocketAddr::new(external.parse().unwrap(), DEFAULT_EXONUM_LISTEN_PORT).to_string(),
+                SocketAddr::new("0.0.0.0".parse().unwrap(), DEFAULT_EXONUM_LISTEN_PORT)
+            )
+        );
+
+        let external = "2001:db8::1";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (
+                SocketAddr::new(external.parse().unwrap(), DEFAULT_EXONUM_LISTEN_PORT).to_string(),
+                SocketAddr::new("::".parse().unwrap(), DEFAULT_EXONUM_LISTEN_PORT)
+            )
+        );
+
+        let external = "[2001:db8::1]:1234";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (external.to_string(), "[::]:1234".parse().unwrap())
+        );
+
+        let external = "localhost";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (
+                format!("{}:{}", external, DEFAULT_EXONUM_LISTEN_PORT),
+                SocketAddr::new("0.0.0.0".parse().unwrap(), DEFAULT_EXONUM_LISTEN_PORT)
+            )
+        );
+
+        let external = "localhost:1234";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (
+                external.to_string(),
+                SocketAddr::new("0.0.0.0".parse().unwrap(), 1234)
+            )
+        );
+
+        let external = "127.0.0.1:1234";
+        let listen = "1.2.3.4:5678";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        ctx.set_arg(LISTEN_ADDRESS, listen.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (external.to_string(), listen.parse().unwrap())
+        );
+
+        let external = "127.0.0.1:1234";
+        let listen = "1.2.3.4:5678";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        ctx.set_arg(LISTEN_ADDRESS, listen.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (external.to_string(), listen.parse().unwrap())
+        );
+
+        let external = "example.com";
+        let listen = "[2001:db8::1]:5678";
+        ctx.set_arg(PEER_ADDRESS, external.to_string());
+        ctx.set_arg(LISTEN_ADDRESS, listen.to_string());
+        assert_eq!(
+            GenerateNodeConfig::addresses(&ctx),
+            (
+                format!("{}:{}", external, DEFAULT_EXONUM_LISTEN_PORT),
+                listen.parse().unwrap()
+            )
+        );
+    }
```

Comment:
I believe it would be better to test `GenerateNodeConfig::addresses()` in isolation and split this one test into several with descriptive names. But the current approach is fine too if no one else objects.

Reply:
Who do you mean? 
Because I think that `GenerateNodeConfig::addresses()` is tested in isolation here. Only call to this function is tested.

Reply:
Oh. I'm terribly sorry. I misread the diff :(